### DISCUSSION TITLE
Best way to train Node classification on WebKB datasets

### DISCUSSION THREAD

**Discussion by AdarshMJ** [pyg-team/pytorch_geometric#6080] [Q&A] [answered]
Hi @EdisonLeeeee @rusty1s I am trying to train a simple GCN on the WebKB dataset for node classification -

```Python
INFO:root:Dataset: cornell()
Dataset: cornell():
====================
Number of graphs: 1
Number of features: 1703
Number of classes: 5

Data(x=[183, 1703], edge_index=[2, 298], y=[183], train_mask=[183, 10], val_mask=[183, 10], test_mask=[183, 10])
=============================================================
Number of nodes: 183
Number of edges: 298
Average node degree: 1.63
Has isolated nodes: False
Has self-loops: True
Is undirected: False
```

Since it has 10 random splits, I am currently optimising my hyper-parameters on the val_mask, that is, for each random split, take average over 5 random seeds run for 100 epochs and I later use test_mask to report the test accuracy. Is this the right way to go about for training them? Or even better what would be the best way to conduct experiments on these datasets? Any suggestions? Thank you for your time!

**Comment by EdisonLeeeee** [comment, marked as answer]
You might refer to [Geom-GCN: Geometric Graph Convolutional Networks](https://arxiv.org/abs/2002.05287) and the [official code](https://github.com/graphdml-uiuc-jlu/geom-gcn), where authors report the average performance of all models on the test sets over 10 random splits with the same hyper-parameter setting.

**Reply by AdarshMJ** [reply]
Hi thanks for the suggestion..  I looked it up. I am having a bit of trouble understanding the train/test/val masks. The paper you referred to says -

` For all graph datasets, we randomly split nodes of each class into 60%, 20%, and 20% for training, validation and testing. With the hyper-parameter setting, we report the average performance of all models on the test sets over 10 random splits.`

Is there a way to split the number of nodes I use for train/test/val?

```Python
Data(x=[183, 1703], edge_index=[2, 298], y=[183], train_mask=[183, 10], val_mask=[183, 10], test_mask=[183, 10])
```

**Reply by AdarshMJ** [reply]
Also this is interesting, originally the graph looked like this -

```Python
Data(x=[183, 1703], edge_index=[2, 298], y=[183], train_mask=[183, 10], val_mask=[183, 10], test_mask=[183, 10])
```
If I print the test_mask -

```Python
print(data.test_mask)

tensor([[ True, False, False,  ..., False, False, False],
        [False, False, False,  ...,  True, False,  True],
        [False, False, False,  ..., False,  True, False],
        ...,
        [False,  True,  True,  ...,  True, False, False],
        [False, False, False,  ..., False,  True, False],
        [False,  True, False,  ..., False,  True, False]])
```

Now if I apply RandomNodeSplit -

```python
from torch_geometric.transforms import RandomNodeSplit
transform = RandomNodeSplit(split='random', num_splits = 10)
data  = transform(data)
print(data)
```
and print the test_mask, I get -

```python
tensor([[False, False, False,  ..., False, False, False],
        [False, False, False,  ..., False, False, False],
        [False, False, False,  ..., False, False, False],
        ...,
        [False, False, False,  ..., False, False, False],
        [False, False, False,  ..., False, False, False],
        [False, False, False,  ..., False, False, False]])

```

All of them are False? And If I try testing on this it gives me division by 0 error. Am I doing something wrong with the splits?

**Reply by rusty1s** [reply]
What does `data.test_mask.sum(dim=0)` return?

**Reply by AdarshMJ** [reply]
It returns all zeros -

`tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])`

But if I were to include this in the transform -

```python
from torch_geometric.transforms import RandomNodeSplit
transform = RandomNodeSplit(split='random', num_splits = 10, num_train_per_class=60, num_val=20, num_test=10)
data  = transform(data)
print(data)
```
The train,val and test, then I get -

```python
print(data.test_mask.sum(dim=0))
tensor([2, 2, 2, 2, 2, 2, 2, 2, 2, 2])
```

So num_train_per_class = 60 means since there are 5 classes, the training set will have 300 nodes? But I only have 183 nodes in total, so this split is wrong?

**Reply by rusty1s** [reply]
Yeah, something like
```python
RandomNodeSplit(split="train_rest", num_val=0.1, num_test=0.1)
```
should fix this.

**Reply by AdarshMJ** [reply]
But I can also set the values for num_train_class = 30, num_val = 15, num_test=18 this way right? As long as I am keeping the total number of nodes in mind?

**Reply by rusty1s** [reply]
Yes, but the actual amount of validation and test nodes might be lower if you set `num_train_class` to a large value.

**Reply by AdarshMJ** [reply]
Oh okay.. thank you. 
I am using 10 random splits and doing this for Cora, since it has 2708 nodes.

`transform = RandomNodeSplit(split='random', num_splits = 10, num_train_per_class=250, num_val=500, num_test=458)`

**Reply by rusty1s** [reply]
Yeah, that should probably work.

**Reply by AdarshMJ** [reply]
Okay. Thank you so much!

**Reply by AdarshMJ** [reply]
Also one last question, sorry. So I can do hyperparameter tuning on the val_mask and then use the best hyper parameters to finally test on the test_mask and report the results right?

**Reply by EdisonLeeeee** [reply]
Exactly :)

**Reply by AdarshMJ** [reply]
Thank you so much :) PyG is an awesome library and thank you for maintaining it so well and also answering all these queries!
